[ESP32] Add camera example for ESP32 platform

sayondeep opened the pull request:
#### Summary
Add Camera example app for ESP32 platform.
The example is supported on ESP32-P4 Function EV Board

#### Testing
CI

sayondeep committed:
Update kvs webrtc sdk submodule

sayondeep committed:
Camera Example on ESP32 Function EV Board

sayondeep committed:
Update camera-app/esp32 data model

sayondeep committed:
Refactoring

sayondeep committed:
Restyling

sayondeep committed:
Address review comments

sayondeep committed:
Merge branch 'master' into pr/camera_example

sayondeep committed:
Fix ESP32_QEMU failure in CI

dhrishi commented on examples/camera-app/esp32/main/Kconfig.projbuild:
```diff
@@ -0,0 +1,129 @@
+        default DEVICE_TYPE_ESP32_S3_DEVKITM if IDF_TARGET_ESP32S3
+        default DEVICE_TYPE_ESP32_C2_DEVKITM if IDF_TARGET_ESP32C2
+        default DEVICE_TYPE_ESP32_C6_DEVKITC if IDF_TARGET_ESP32C6
+        default DEVICE_TYPE_ESP32_H2_DEVKITC if IDF_TARGET_ESP32H2
```
Are all of these valid here? If not, can we just remove the unwanted things

shubhamdp commented on examples/camera-app/esp32/README.md:
We should add a pointers to camera-controller-app, and esp-webrtc/esp_port/examples/streaming_only, and specific esp-idf commit/branch that should be used.

shubhamdp commented on examples/camera-app/esp32/sdkconfig.defaults:
```diff
@@ -0,0 +1,207 @@
+#
+#    Copyright (c) 2021 Project CHIP Authors
```
```suggestion
#    Copyright (c) 2025 Project CHIP Authors
```

dhrishi commented on examples/camera-app/esp32/sdkconfig.defaults.esp32c6:
```diff
@@ -0,0 +1,36 @@
+CONFIG_FREERTOS_GENERATE_RUN_TIME_STATS=y
+
+# 6K is sufficient for signalling_only
+CONFIG_MAIN_TASK_STACK_SIZE=4000
```
Is the change done? 6k and 4000 inconsistency?

shubhamdp commented on examples/camera-app/esp32/sdkconfig.defaults:
```diff
@@ -0,0 +1,207 @@
+#
+CONFIG_ESP_WIFI_SSID="myssid"
+CONFIG_ESP_WIFI_PASSWORD="mypassword"
+CONFIG_ESP_MAXIMUM_RETRY=5
```
why are these needed?

shubhamdp commented on examples/camera-app/esp32/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+#
+#    Copyright (c) 2021-2023 Project CHIP Authors
```
```suggestion
#    Copyright (c) 2025 Project CHIP Authors
```

dhrishi commented on examples/camera-app/esp32/main/webrtc/webrtc-abstract.h:
```diff
@@ -0,0 +1,58 @@
+#pragma once
+
```
@vikramdattu Please review the webrtc related changes here

vikramdattu commented on examples/camera-app/esp32/main/clusters/webrtc-provider-manager.cpp:
```diff
@@ -0,0 +1,1096 @@
+#include "webrtc-provider-manager.h"
```
missing license header. do check for other instances as well

shubhamdp commented on examples/camera-app/esp32/main/Kconfig.projbuild:
```diff
@@ -0,0 +1,129 @@
+        string "CHIP Project Configuration file"
+        default "main/include/CHIPProjectConfig.h"
+
+    choice
```
even this and the `RENDEZVOUS_MODE` as we generally generate the onboarding code outside of the ðevice

shubhamdp commented on examples/camera-app/esp32/main/Kconfig.projbuild:
```diff
@@ -0,0 +1,129 @@
+
+endmenu
+
+menu "PW RPC Debug channel"
```
I think we are not using PW RPC in this example, can this be removed?

vikramdattu commented on examples/camera-app/esp32/main/clusters/webrtc-provider-manager.cpp:
```diff
@@ -0,0 +1,1096 @@
+    std::string localSdp = sdp;
+    ChipLogProgress(Camera, "Local Description (%s):", typeStr);
+    ChipLogProgress(Camera, "%s", localSdp.c_str());
+    printf("Size of SDP: %zu\n", localSdp.size());
```
Use ChipLog instead or merge in above local SDP print

shubhamdp commented on examples/camera-app/esp32/main/idf_component.yml:
```diff
@@ -0,0 +1,22 @@
+dependencies:
+    network_coprocessor:
+        version: "*"
```
I hope the `*` won't keep breaking the builds

vikramdattu commented on examples/camera-app/esp32/main/clusters/webrtc-provider-manager.cpp:
```diff
@@ -0,0 +1,1096 @@
+        ScheduleOfferSend(sessionId);
+        break;
+    case WebrtcTransport::State::SendingAnswer:
+        // sdp_mem_dump();
```
please remove the commented code

vikramdattu commented on examples/camera-app/esp32/main/main.cpp:
```diff
@@ -0,0 +1,227 @@
+}
+
+#ifdef CONFIG_SLAVE_LWIP_ENABLED
+static void create_slave_sta_netif(uint8_t dhcp_at_slave)
```
Should not be needed anymore. Please refer to the rainmaker example: https://github.com/espressif/esp-rainmaker/blob/master/examples/camera/split_mode/rmaker_split_camera/main/app_main.c#L223

vikramdattu commented on examples/camera-app/esp32/main/webrtc/webrtc-kvs_esp_port_utils.cpp:
```diff
@@ -0,0 +1,340 @@
+    jsmn_parser parser;
+    jsmntok_t tokens[64];
+    int ret;
+
```
should have sanity checks on json and sdp_buf

vikramdattu commented on examples/camera-app/esp32/main/webrtc/webrtc-kvs_esp_port_utils.cpp:
```diff
@@ -0,0 +1,340 @@
+    // handle message
+    printf("Received Message from P4-Streamer: \n%.*s\n", len, (char *) data);
+
+    signaling_msg_t * msg = new signaling_msg_t;
```
Add sanity check here.

vikramdattu reviewed (commented):
@sayondeep added first round of review.

shubhamdp commented on examples/camera-app/esp32/main/camera-device.cpp:
```diff
@@ -0,0 +1,630 @@
+#include <iostream>
+#include <lib/support/logging/CHIPLogging.h>
+#include <limits.h>
+#include <sys/ioctl.h>
```
this is also unused

sayondeep committed:
Add Button callback for putting into Standby mode

sayondeep commented on examples/camera-app/esp32/main/Button.cpp:
```diff
@@ -0,0 +1,111 @@
+    // Get the button index of the expired timer and call button event Handler.
+    uint32_t gpio_num = (uint32_t) pvTimerGetTimerID(xTimer);
+    GetAppTask().ButtonEventHandler(gpio_num, APP_BUTTON_PRESSED);
+}
```
Added support for putting camera into standby mode using buttons.

sayondeep committed:
Add instruction in Readme

wqx6 commented on examples/camera-app/esp32/main/Button.cpp:
```diff
@@ -0,0 +1,66 @@
+     }
+ }
+
+ void Button::Init()
```
Are we using any buttons on C6 in this example? If no, please remove this file and the header.

sayondeep committed:
Address Review comments

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by prettier-markdown

sayondeep committed:
Add camera-app/esp32 in CI build

restyled-commits committed:
Restyled by shellharden

sayondeep committed:
Address review comments

restyled-commits committed:
Restyled by clang-format

sayondeep committed:
Merge branch 'master' into pr/camera_example

sayondeep committed:
Update Readme

sayondeep commented on examples/camera-app/esp32/main/idf_component.yml:
```diff
@@ -0,0 +1,22 @@
+    esp_webrtc_utils:
+        path: ${KVS_SDK_PATH}/esp_port/components/esp_webrtc_utils
+        version: "*"
+
```
Check and Remove this component

sayondeep commented on examples/camera-app/esp32/main/webrtc/webrtc-kvs_esp_port.cpp:
```diff
@@ -0,0 +1,184 @@
+    if (message == nullptr)
+    {
+        ChipLogError(Camera, "AddRemoteCandidate: failed to allocate signaling_msg_t");
+        return;
```
cleanup candidate_json befor return

sayondeep committed:
Update README.md

sayondeep committed:
Update README.md

sayondeep committed:
Update README.md

restyled-commits committed:
Restyled by prettier-markdown

sayondeep committed:
Merge remote-tracking branch 'origin/master' into pr/camera_example

sayondeep committed:
camera data model fix and AVSUM cluster fix

sayondeep committed:
Update avsm cluster delegate

sayondeep committed:
Update webrtc cluster delegate

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

sayondeep committed:
Update example for code driven clusters

sayondeep committed:
Merge branch 'master' into pr/camera_example

sayondeep committed:
Add cluster init callbacks

sayondeep committed:
Update esp32 camera-app.matter

restyled-commits committed:
Restyled by clang-format

sayondeep committed:
Merge branch 'master' into pr/camera_example

sayondeep committed:
Fix Camera clusters for TC-IDM tests

sayondeep committed:
update esp camera data model

sayondeep committed:
Fix data model versions

sayondeep committed:
Add C5 support

sayondeep committed:
Add snapshot support

sayondeep committed:
Save snapshot

sayondeep committed:
SDP length allocation in heap

sayondeep committed:
Merge branch 'master' into pr/camera_example

restyled-commits committed:
Restyled by clang-format

sayondeep committed:
updated snapshot quality

sayondeep committed:
Merge branch 'master' into pr/camera_example

sayondeep committed:
third_party: Update amazon kvs webrtc sdk

sayondeep committed:
Merge remote-tracking branch 'upstream/master' into pr/camera_example

sayondeep committed:
camera-app/esp32: Fix data model

sayondeep committed:
camera-app/esp32: Include uriparser in build

sayondeep committed:
Merge branch 'master' into pr/camera_example

sayondeep committed:
Move to beta-reference-esp-port branch of KVS SDK

sayondeep committed:
third_party: Update amazon kvs webrtc sdk to support snapshot

sayondeep committed:
Merge branch 'master' of https://github.com/project-chip/connectedhomeip into pr/camera_example

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

sayondeep committed:
examples/camera-app/esp32: Update data model

vikramdattu commented on examples/camera-app/esp32/sdkconfig.defaults.esp32c5:
```diff
@@ -0,0 +1,112 @@
+CONFIG_IDF_TARGET="esp32c6"
```
Please change this to c5

sayondeep replied:
made the change

vikramdattu commented on examples/camera-app/esp32/README.md:
```diff
@@ -0,0 +1,135 @@
+
+### Prerequisites
+
+-   IDF version: v5.4.1
```
This is 5.5 now?

sayondeep replied:
yes, made the change to 5.5.4

vikramdattu commented on examples/camera-app/esp32/main/clusters/webrtc-provider-manager.cpp:
```diff
@@ -0,0 +1,1144 @@
+        // and then sends an End command with OutOfResources reason.
+        ScheduleEndSend(args.sessionId);
+
+        return CHIP_NO_ERROR;
```
Use MEM related error instead?

sayondeep replied:
Its not actually memory related error on the CHIP side.
It is max streaming sessions supported by the application. We are logging appropriate Message.

vikramdattu commented on examples/camera-app/esp32/main/clusters/webrtc-provider-manager.cpp:
```diff
@@ -0,0 +1,1144 @@
+{
+    ChipLogProgress(Camera, "Initializing WebRTC PeerConnection");
+    // Register our handler for signaling messages
+    webrtc_bridge_register_handler((webrtc_bridge_msg_cb_t) &webrtc_bridge_message_received_cb);
```
typecasting hides if webrtc_bridge_message_received_cb signature does not follow webrtc_bridge_msg_cb_t

sayondeep replied:
Made correction in the signature and removed typecasting

vikramdattu commented on examples/camera-app/esp32/main/clusters/webrtc-provider-manager.cpp:
```diff
@@ -0,0 +1,1144 @@
+
+    outDeferredOffer = deferred_offer;
+
+    ChipLogError(Camera, "Deferred Offer is set to: %d", outDeferredOffer);
```
Should be Info/Debug log instead

vikramdattu commented on examples/camera-app/esp32/main/webrtc/webrtc-kvs_esp_port.cpp:
```diff
@@ -0,0 +1,189 @@
+{
+    // handles SDP Offer received from webrtc requestor.
+    // Send SDP to KVSWebRTCManager.
+    char * sdp_json = (char *) heap_caps_malloc_prefer(gSDPLength, 2, MALLOC_CAP_8BIT | MALLOC_CAP_SPIRAM, MALLOC_CAP_INTERNAL);
```
Can we calculate the actual required size instead

sayondeep replied:
Yes made the changes, removed fixed size alloc.

vikramdattu commented on examples/camera-app/esp32/main/webrtc/webrtc-kvs_esp_port_utils.cpp:
```diff
@@ -0,0 +1,382 @@
+    ret = jsmn_parse(&parser, json, strlen(json), tokens, sizeof(tokens) / sizeof(tokens[0]));
+    if (ret < 0)
+    {
+        printf("Failed to parse JSON: %d\n", ret);
```
Can we use CHIP logs OR ESP_LOG instead?

sayondeep replied:
Made the change

vikramdattu commented on examples/camera-app/esp32/main/webrtc/webrtc-kvs_esp_port_utils.cpp:
```diff
@@ -0,0 +1,382 @@
+            int len = tokens[i + 1].end - tokens[i + 1].start;
+            if (len >= sdp_buf_len)
+                len = sdp_buf_len - 1;
+            strncpy(sdp_buf, json + tokens[i + 1].start, len);
```
This is a potential overflow if the sdp_buf size is smaller than len

sayondeep replied:
Fixed

vikramdattu commented on examples/camera-app/esp32/main/webrtc/webrtc-kvs_esp_port_utils.cpp:
```diff
@@ -0,0 +1,382 @@
+void webrtc_bridge_message_received_cb(void * data, int len)
+{
+    // handle message
+    printf("Received Message from P4-Streamer: \n%.*s\n", len, (char *) data);
```
Change to ESP_LOG or Chip log. Check other instances too

sayondeep replied:
Checked and removed printf statements

vikramdattu reviewed (commented):
@sayondeep added a few comments. Please check

sayondeep commented on examples/camera-app/esp32/main/Kconfig.projbuild:
```diff
@@ -0,0 +1,129 @@
+
+menu "Demo"
+
+    choice
```
Added support for esp32c5 as well

sayondeep committed:
Fix logging and memory issues on esp32c5

sayondeep committed:
Merge branch 'master' of https://github.com/project-chip/connectedhomeip into pr/camera_example

sayondeep committed:
[camera-app][esp32] Size WebRTC signaling buffers exactly; fix unsafe allocs

The ESP32 WebRTC port allocated fixed-size buffers (gSDPLength,
gCandidateLength) for SDP/candidate JSON and used unchecked malloc +
sprintf, risking truncation on large SDPs and overflow if the size
constants ever drifted.

- SetRemoteDescription / AddRemoteCandidate: choose the format string
  first, size the buffer via a snprintf() length probe, then write with
  snprintf(). Allocation failure is now checked; the early-return path in
  AddRemoteCandidate no longer leaks candidate_json.
- webrtc_bridge_message_received_cb: size the extract buffer from
  msg->payloadLen + 1 (the extracted value is a substring of the payload)
  and pass that size to extract_sdp/extract_candidate. Added a guard for
  empty/null payloads.
- Removed the now-unused gCandidateLength/gSDPLength globals.

Also align webrtc_bridge_message_received_cb's signature with
webrtc_bridge_msg_cb_t (const void *) and drop the C-style cast at the
registration site so the compiler enforces the callback signature.

sayondeep committed:
[camera-app][esp32] Merge JSON extractors, use ChipLog over printf

Consolidate the WebRTC signaling JSON parsing and clean up logging in
the ESP32 camera app:

- Merge extract_sdp() and extract_candidate() into a single
  extract_json_field(json, key, out_buf, out_buf_len) helper, and route
  OFFER/ANSWER (key "sdp") and ICE_CANDIDATE (key "candidate") through it.
- Fix out-of-bounds token access: the loop now bounds tokens[i + 1]
  (i + 1 < ret) instead of reading past the parsed set on a trailing key.
- Require an exact key-length match before strncmp so a short key like
  "sd" can no longer match "sdp".
- Clamp the copy length entirely in size_t space (memcpy + explicit null
  terminator), avoiding the prior signed/unsigned comparison.
- Replace printf() calls with ChipLogProgress/ChipLogError, and remove the
  duplicate "Set SDP Answer/Candidate to WebRTCProviderManager" logs that
  fired even when the transport was not found.
- HandleProvideOffer: log under ChipLogProgress(Camera) instead of
  ChipLogError(Zcl) for a normal-path message.

sayondeep committed:
examples/camera-app/esp32:Update data model

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

sayondeep committed:
Merge branch 'master' into pr/camera_example
